cpp/src/arrow/compute/registry.h

```diff
@@ -47,35 +47,64 @@ class ARROW_EXPORT FunctionRegistry {
  public:
   ~FunctionRegistry();
 
-  /// \brief Construct a new registry. Most users only need to use the global
-  /// registry
+  /// \brief Construct a new registry.
+  ///
+  /// Most users only need to use the global registry.
   static std::unique_ptr<FunctionRegistry> Make();
 
-  /// \brief Add a new function to the registry. Returns Status::KeyError if a
-  /// function with the same name is already registered
+  /// \brief Construct a new nested registry with the given parent.
+  ///
+  /// Most users only need to use the global registry. The returned registry never changes
+  /// its parent, even when an operation allows overwritting.
+  static std::unique_ptr<FunctionRegistry> Make(FunctionRegistry* parent);
+
+  /// \brief Check whether a new function can be added to the registry.
+  ///
+  /// \returns Status::KeyError if a function with the same name is already registered.
+  Status CanAddFunction(std::shared_ptr<Function> function, bool allow_overwrite = false);
```

Comment:
nit: take `const std::shared_ptr<Function>&` since this is a check? Also, I would expect this to return `bool` and be declared `const`

Reply:
Both `const` should be fine, so I'll try that. My reason for using `Status`  as the return type is that the caller gets an explanation when not OK.

Reply:
Turns out changing to `const` causes a compilation error when the implementation calls `DoAddFunction`, which is designed to be called from both `AddFunction` and `CanAddFunction`. I wouldn't like to resolve this by using a `const_cast` or by creating `const` and non-`const` versions of `DoAddFunction` (perhaps with different names). Moreover, such two versions involve a non-reentrant mutex acquisition that would make it annoying to avoid code duplication.

Reply:
I suppose why does **Can**AddFunction call **Do**AddFunction? Unless you meant to make it **Try**AddFunction or something (I think that is my confusion). In the latter case then the existing signature is fine.

Reply:
`DoAddFunction` either does the addition or dry-runs it, depending on the Boolean flag argument. I guess `Do` isn't ideal to describe this but `Try` would not be right. This dry-run feature is why `CanAddFunction` and `AddFunction` are both implemented using a common function and differ only by the value of a flag.

Reply:
Ah…here I had thought the `allow_overwrite` was meant for the user not for the internal implementation. Can't we factor out an explicit check instead of relying on a flag?

Reply:
Sorry, looks like I confused you. The flag I meant is [`bool add` of `DoAddFunction`](https://github.com/apache/arrow/pull/13375/files#diff-703810f407203fb49563d6b8dc719052ee0fe9828559dc29bc667040cf3ebbbaR158) - compare the [invocation in `CanAddFunction`](https://github.com/apache/arrow/pull/13375/files#diff-703810f407203fb49563d6b8dc719052ee0fe9828559dc29bc667040cf3ebbbaR45) with [that in `AddFunction`](https://github.com/apache/arrow/pull/13375/files#diff-703810f407203fb49563d6b8dc719052ee0fe9828559dc29bc667040cf3ebbbaR52).

Reply:
Ok. I suppose constness isn't a big deal here. The API seems a little complicated for what's going on but I suppose having the lock makes it hard to factor things out conveniently.

Reply:
OK, I'll keep this code.